# CRISPResso2parser.py

These Python parser scripts summarize and genotype CRISPResso2 analysed clones including those edited with multiple HDRs.

These scripts are based on previous repositories of MATLAB code. They extract and join relevant data from the results folders\
generated by CRISResso2 per each sample. (K. Clement et al. Nat. Biotechnology 2019)\ (https://github.com/pinellolab/CRISPResso2) (http://crispresso2.pinellolab.org/).

# REQUERIMENTS
 \
Python 2.7 with openpyxl module.
 \
# GENERAL WORKFLOW
 \
Copy the CRISPYparser....py file in your folder containing CRISPResso_on_··· subfolders generated by CRISPResso2.\
 \
Open a Terminal window or python development enviroment such as Spyder.\
 \
Run the script by typing "python CRISPResso2parser...py" in a Terminal window or clicking in run button.\
 \
Introduce your input data asked by the script in the Terminal or enviroment command window.\
 \
The script will generate an excel (.xlsx) file with your CRISPResso2 results summary.\
\
# USAGE OF THE DIFFERENT PARSER SCRIPTS\
 \
######

  The percentage of the 2 top reads from each fastq file. These data will be referred as Alleles percentages.
  Determine the genotype compatible with the reads pattern of each fastq file. Genotype determination is based on Alleles
  percentages and the introduced Purity_Percentage.

Finally, the script attachs these data to the CSV result file and generates an excel file (.xlsx) in the working directory.

# ADVANTAGES

The excel file generated by CRISPYparser.m allows easier visualization of data coming from CRIS.py analysis.
Furthermore, CRISPYparser.m identifies pure clones and determines their zygosity. Concurrent visualization of 
alleles and CRIS.py analysed sequences percentages allows the identification of more detailed genotypes by
conditional formulations in excel.
